### DISCUSSION TITLE
Request Types of TypedDict cannot be parsed for Swagger / OpenAPI Specifications when Required attribute is used.

### DISCUSSION THREAD

**Discussion by jperuggia** [fastapi/fastapi#11952] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Removing the future imports allows the code to work as expected.
from __future__ import annotations

from fastapi import FastAPI 
from typing_extensions import Required, NotRequired, Literal, TypedDict

# IF you want to use OpenAI class as example, but not required.
# https://github.com/openai/openai-python/blob/da48e4cac78d1d4ac749e2aa5cfd619fde1e6c68/src/openai/types/chat/chat_completion_system_message_param.py
# from openai.types.chat.chat_completion_system_message_param import ChatCompletionSystemMessageParam

app = FastAPI()

class RequiredTypedDictExample(TypedDict, total=False):
    name: Required[str]= "This is required"
    age: int

class NonRequiredTypedDictExample(TypedDict, total=False):
    name: NotRequired[str]
    age: int

@app.post("/RequiredTypeDict")
async def something_to_post(request: RequiredTypedDictExample):
    return {"message": "Hello World"}

@app.post("/NonRequiredTypedDictExample")
async def something_to_post_2(request: NonRequiredTypedDictExample):
    return {"message": "Hello World"}

# @app.post("/OpenAITestCase")
# async def something_to_post_3(request: ChatCompletionSystemMessageParam):
#     return {"message": "Hello World"}
```

### Description

Running a simple example with a POST endpoint that expects a request parameter defined by a TypedDict class results in an error within Pydantic about 
```
TypeError: 'pydantic_core._pydantic_core.PydanticUndefinedType' object is not subscriptable
```

Trying to dig into the issue more lead me to see that the `Required` attribute / import is being redefined as PydanticUndefined in the [file here](https://github.com/fastapi/fastapi/blob/af1a07052a713da96c5c75c9548203c77d0b6fd6/fastapi/_compat.py#L73)

When importing types defined by other modules / classes it isn't always possible to change. 

If you run the above, but remove the `from __future__ import annotations`, the code runs and performs as expected. 

Not sure why the futures import causes issues here, but with this I am unable to use classes / datatypes defined in other modules that make use of this python feature. 

### Operating System

Linux

### FastAPI Version

0.112.0

### Pydantic Version

2.8.2

### Python Version

Python 3.12.2

**Comment by pachewise** [comment]
https://github.com/fastapi/fastapi/pull/12197

**Comment by YuriiMotov** [comment]
Error is reproducible with Pydantic 2.8.2, but everything is Ok with Pydantic 2.9+.
So, seems that it was fixed on Pydantic's side.

Let's close this for now to keep the discussions clean.
If you still need help, feel free to let us know.
